### DISCUSSION TITLE
Nvidia Tesla P4 GPU stuck in P0 state

### DISCUSSION THREAD

**Discussion by obones** [blakeblackshear/frigate#6850] [Ask A Question]
Hello,

I'm just starting to explore frigate and it works well "out of the box" despite having high CPU usage by ffmpeg.
Luckily, I a Nvidia Tesla P4 GPU card in the system so I configured frigate to tell ffmpeg to use it like so:

```
ffmpeg:
  hwaccel_args: preset-nvidia-h264
```

It works quite well, CPU usage is down to almost nothing but I have another issue that leads to GPU shut down because of overheating. 
`nvidia-smi` tells me that the card keeps running in P0 power state, drawing 25W and never moving back to P2 as shown in the frigate [documentation](https://docs.frigate.video/configuration/hardware_acceleration#nvidia-gpus):

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.41.03              Driver Version: 530.41.03    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla P4                        Off| 00000000:01:00.0 Off |                    0 |
| N/A   46C    P0               24W /  75W|    330MiB /  7680MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A    811348      C   ffmpeg                                      328MiB |
+---------------------------------------------------------------------------------------+
```

Does any of you have any idea what could be causing this? I'm quite positive this is an issue with the GPU configuration rather than with frigate but I could not find any solution and hope someone may have hints for a solution.

**Comment by NickM-27** [comment]
I'd suggest trying an updated driver if one is available. 

I also wouldn't necessarily take that part of the frigate documentation as what it should look like, different GPUs will behave differently.

**Comment by NickM-27** [comment]
You can also use nvidia smi on the host to set the p state

**Comment by hardwareadictos** [comment]
I have the same issue with a P4, latest grid drivers:, GPU is attatched via vgpu to another 3 vms and it only reaches P8 state when i stop Frigate container, using   hwaccel_args: preset-nvidia-h264 also, happens with   hwaccel_args: preset-nvidia-h265 too.

Would be nice to know which behaviour are other nvidia users experiencing.

**Comment by hardwareadictos** [comment]
Okey, so some months ago the issue still persists, to go deeper on the issue i have to clarify that i always had my Guest/vGPU drivers updated to latest versions, the issue seems to be with hardware acceleration:

![imagen](https://github.com/user-attachments/assets/c22cccb0-a756-4a3e-8c2e-fbe80b4ada77)

I hace 3 cameras but i **DON'T HAVE** any hwaccel profile on my config. I only enabled tensorrt for object detection which barelly uses 1% of the GPU. Why is Frigate using my P4 for HW Accelleration when i dont have any preset enabled??

![imagen](https://github.com/user-attachments/assets/1956b0bb-7989-4e60-b68e-9cee4c114391)

My P4 only leaves P0 state when i stop my Frigate container.

**Reply by NickM-27** [reply]
Hardware acceleration is automatically used if it is detected

**Reply by hardwareadictos** [reply]
Is it any way to disable hardware accelleration? On my P4 tensorrt barely goes down from P8 state, i don't need any hardware transcoding at all, i only need AI for object detection. HW accelleration is causing my GPU to go to P0 state making power consumtion higher than needeed, and i don't want my GPU to be at 70C+ always for anything i don't need :(

**Reply by NickM-27** [reply]
yes, just manually disable it

```
ffmpeg:
  hwacce_args: []
```

**Reply by hardwareadictos** [reply]
Thanks a lot @NickM-27 , let me test!

**Reply by hardwareadictos** [reply]
Well, with ** hwaccell_args: []** now my P4 downclocks and uses up to P8 state, but it still uses P0 when detection is active and tensor comes to scene. Maybe that's now normal

**Comment by obones** [comment]
FWIW, I "fixed" it by replacing the P4 with a RTX3060 as I was using other projects that exhibited the same behavior. Now it sits at P2 most of the time which is what I expected.

**Reply by hardwareadictos** [reply]
There is a real reason for that: As RTX3060 is likely 150% more powerful and has more moderen decorders/encoders it probably doesn 't need to use P0 for transcoding. The real issue here is that frigate is directly activating transcoding by default when it founds a GPU on the container.

**Reply by obones** [reply]
I'm not so sure about that, the other project that exhibited the same behavior is Willow and it doesn't use any video decoder/encoder, it's using the GPU to do STT/TTS inference

**Comment by leccelecce** [comment]
Out of curiosity, presumably you've measured the total power draw of your system at the wall (i.e. not just the GPU's power consumption) and confirmed you're not just shifting the decode power consumption to the CPU? 70C is not that hot for a GPU.

To be honest I imagine you'll probably find it *is* saving you to run it the way you're trying (i.e. the minimum active cost of the GPU could be relatively high) but curious - I am on a journey of power consumption savings too!

**Reply by leccelecce** [reply]
My other point would be, if object detection only uses your GPU 1%, do you really need a GPU at all...

**Reply by hardwareadictos** [reply]
Well TBO i'm not afraid of the power consumption as this GPU is on a Rack which consumes about 400W per hour, but i don't want extra consumptions for nothing, that's why i asked to disable HW accelleration. Anyways as i am going to explain up there, hwaccel wasn't the issue.
